pgo: add auth connections stress workload

nuivall opened the pull request:
This series improves the pgo workloads by enabling authentication and authorization and adding new
stress scenarios.

- Enables auth in training clusters
  All training workloads now run with auth enabled, following best
  practices and avoiding config proliferation.

- Adds auth connections stress workload
  Introduces a workload that uses derived roles and permissions, stressing auth
  code paths while also creating a new connection per request to exercise
  server transport handling.

- Enables counters workload
  The counters workload is re-enabled without introducing extra dependencies on
  cqlsh. Instead, a lightweight exec_cql.py wrapper (shared with the auth
  workload) handles preparation statements.

Backport: no, it's not a bug fix

----------------------------------------------------------
Performance results for auth PGO there seems to be no difference, or to small to measure:

scylladb pgo_auth ≡ ◦ ⤖ python3 ./pgo/auth_conns_stress.py localhost cassandra cassandra 10000 100 &
scylladb pgo_auth ≡ ◦ ⤖ perf stat -e instructions --timeout 5000 -p 51591

on both before and after instructions counter varies from 179,818,558,011 to 180,664,528,198.

----------------------------------------------------------
Performance results for counters PGO is notably improved with write workload 16-22% and read 4-5%:

scylladb pgo_auth ≡ ◦ ⤖ ./scylla_master perf-simple-query 2> /dev/null --counters --write
random-seed=3839439576
enable-cache=1
Running test with config: {partitions=10000, concurrency=100, mode=write, query_single_key=no, counters=yes}
Disabling auto compaction
2413435.37 tps (122.1 allocs/op,   8.0 logallocs/op,  14.0 tasks/op,   51167 insns/op,   33157 cycles/op,        0 errors)
2413009.40 tps (122.1 allocs/op,   8.0 logallocs/op,  14.0 tasks/op,   51053 insns/op,   33009 cycles/op,        0 errors)
2403794.31 tps (122.1 allocs/op,   8.0 logallocs/op,  14.0 tasks/op,   50867 insns/op,   32899 cycles/op,        0 errors)
2384572.52 tps (122.1 allocs/op,   8.0 logallocs/op,  14.0 tasks/op,   50562 insns/op,   32811 cycles/op,        0 errors)
2195388.31 tps (122.6 allocs/op,   8.0 logallocs/op,  14.1 tasks/op,   51818 insns/op,   34504 cycles/op,        0 errors)
throughput:
	mean=   2362039.98 standard-deviation=93892.61
	median= 2403794.31 median-absolute-deviation=50969.42
	maximum=2413435.37 minimum=2195388.31
instructions_per_op:
	mean=   51093.44 standard-deviation=465.18
	median= 51052.98 median-absolute-deviation=226.61
	maximum=51818.04 minimum=50562.30
cpu_cycles_per_op:
	mean=   33275.85 standard-deviation=698.65
	median= 33008.58 median-absolute-deviation=377.16
	maximum=34504.13 minimum=32811.18


scylladb pgo_auth ≡ ◦ ⤖ ./scylla_master perf-simple-query 2> /dev/null --counters 
random-seed=1134551638
enable-cache=1
Running test with config: {partitions=10000, concurrency=100, mode=read, query_single_key=no, counters=yes}
Disabling auto compaction
Creating 10000 partitions...
5499534.56 tps ( 73.0 allocs/op,   0.0 logallocs/op,  18.0 tasks/op,   21463 insns/op,   14902 cycles/op,        0 errors)
5478913.87 tps ( 73.0 allocs/op,   0.0 logallocs/op,  18.0 tasks/op,   21385 insns/op,   14839 cycles/op,        0 errors)
5346525.04 tps ( 73.0 allocs/op,   0.0 logallocs/op,  18.1 tasks/op,   21454 insns/op,   15082 cycles/op,        0 errors)
5467947.74 tps ( 73.0 allocs/op,   0.0 logallocs/op,  18.0 tasks/op,   21275 insns/op,   14775 cycles/op,        0 errors)
5454894.98 tps ( 73.0 allocs/op,   0.0 logallocs/op,  18.0 tasks/op,   21250 insns/op,   14766 cycles/op,        0 errors)
throughput:
	mean=   5449563.24 standard-deviation=59878.80
	median= 5467947.74 median-absolute-deviation=29350.63
	maximum=5499534.56 minimum=5346525.04
instructions_per_op:
	mean=   21365.28 standard-deviation=98.95
	median= 21384.65 median-absolute-deviation=90.57
	maximum=21463.17 minimum=21250.33
cpu_cycles_per_op:
	mean=   14872.93 standard-deviation=129.26
	median= 14838.65 median-absolute-deviation=97.52
	maximum=15082.44 minimum=14766.13


scylladb pgo_auth ≡ ◦ ⤖ ./scylla_pgo_counters perf-simple-query 2> /dev/null --counters --write
random-seed=437758611
enable-cache=1
Running test with config: {partitions=10000, concurrency=100, mode=write, query_single_key=no, counters=yes}
Disabling auto compaction
2950968.10 tps (122.1 allocs/op,   8.0 logallocs/op,  14.0 tasks/op,   41540 insns/op,   27097 cycles/op,        0 errors)
2923325.10 tps (122.1 allocs/op,   8.0 logallocs/op,  14.0 tasks/op,   41366 insns/op,   27017 cycles/op,        0 errors)
2928666.67 tps (122.1 allocs/op,   8.0 logallocs/op,  14.0 tasks/op,   41274 insns/op,   26929 cycles/op,        0 errors)
2918378.39 tps (122.1 allocs/op,   8.0 logallocs/op,  14.0 tasks/op,   41165 insns/op,   26880 cycles/op,        0 errors)
2209053.17 tps (128.4 allocs/op,   8.0 logallocs/op,  14.6 tasks/op,   48176 insns/op,   34726 cycles/op,        0 errors)
throughput:
	mean=   2786078.28 standard-deviation=322807.25
	median= 2923325.10 median-absolute-deviation=142588.38
	maximum=2950968.10 minimum=2209053.17
instructions_per_op:
	mean=   42704.41 standard-deviation=3062.05
	median= 41366.40 median-absolute-deviation=1430.69
	maximum=48176.45 minimum=41165.23
cpu_cycles_per_op:
	mean=   28529.92 standard-deviation=3464.99
	median= 27016.51 median-absolute-deviation=1601.02
	maximum=34726.49 minimum=26880.18


scylladb pgo_auth ≡ ◦ ⤖ ./scylla_pgo_counters 2> /dev/null --counters
random-seed=4277130772
enable-cache=1
Running test with config: {partitions=10000, concurrency=100, mode=read, query_single_key=no, counters=yes}
Disabling auto compaction
Creating 10000 partitions...
5691320.62 tps ( 73.0 allocs/op,   0.0 logallocs/op,  18.0 tasks/op,   20656 insns/op,   14279 cycles/op,        0 errors)
5708878.25 tps ( 73.0 allocs/op,   0.0 logallocs/op,  18.0 tasks/op,   20486 insns/op,   14104 cycles/op,        0 errors)
5727060.22 tps ( 73.0 allocs/op,   0.0 logallocs/op,  18.0 tasks/op,   20439 insns/op,   14044 cycles/op,        0 errors)
5700157.92 tps ( 73.0 allocs/op,   0.0 logallocs/op,  18.0 tasks/op,   20416 insns/op,   14054 cycles/op,        0 errors)
5610730.84 tps ( 73.0 allocs/op,   0.0 logallocs/op,  18.0 tasks/op,   20459 insns/op,   14195 cycles/op,        0 errors)
throughput:
	mean=   5687629.57 standard-deviation=44972.99
	median= 5700157.92 median-absolute-deviation=21248.68
	maximum=5727060.22 minimum=5610730.84
instructions_per_op:
	mean=   20491.27 standard-deviation=95.74
	median= 20459.35 median-absolute-deviation=52.13
	maximum=20656.27 minimum=20415.95
cpu_cycles_per_op:
	mean=   14135.25 standard-deviation=100.27
	median= 14104.47 median-absolute-deviation=81.48
	maximum=14278.97 minimum=14043.76

nuivall commented on pgo/conf/counters.cql:
```diff
@@ -0,0 +1,5 @@
+
+CREATE KEYSPACE IF NOT EXISTS counters WITH replication = {'class': 'NetworkTopologyStrategy', 'dc1': '3'};
+
+CREATE TABLE IF NOT EXISTS counters.counter1 (key blob PRIMARY KEY, "C0" counter, "C1" counter, "C2" counter, "C3" counter, "C4" counter);
```
so schema will need to be defined like for si and lwt:
```
# The CQL for creating a keyspace (optional if it already exists)
keyspace_definition: |
  CREATE KEYSPACE ks WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 3} AND TABLETS = {'enabled': false};
```

avikivity commented:
> Adds auth connections stress workload
Introduces a workload that uses derived roles and permissions, stressing auth
code paths while also creating a new connection per request to exercise
server transport handling.

Is this really a hot path? IIRC the derived permissions are cached, so the code to calculate derived permissions isn't touched often.

avikivity commented:
> Backport: we can safely backport to all versions to benefit new training


The question it not if we can, it's if we should.

We're way too backport-happy. We should backport things that are needed, not that can be backported.

nuivall commented:
> > Adds auth connections stress workload
> > Introduces a workload that uses derived roles and permissions, stressing auth
> > code paths while also creating a new connection per request to exercise
> > server transport handling.
> 
> Is this really a hot path? IIRC the derived permissions are cached, so the code to calculate derived permissions isn't touched often.

Yes I think in both cases code paths are not very big, but maybe the cache part can be optimized too. There is also the whole connection management, startup frames parsing, semaphores etc. I don't know how well this optimizes. I'm trying to test before/after, although we don't have full connection storm perf test.

nuivall commented:
> > > > Adds auth connections stress workload
> > > > Introduces a workload that uses derived roles and permissions, stressing auth
> > > > code paths while also creating a new connection per request to exercise
> > > > server transport handling.
> > > 
> > > 
> > > Is this really a hot path? IIRC the derived permissions are cached, so the code to calculate derived permissions isn't touched often.
> > 
> > 
> > Isn't the whole authentication path 'hot' for the 'monday morning' effect where gozillion clients connect to a node that rebooted for example?
> > Unsure about Alternator, btw, where each request is a connection or something?
> 
> yes I think alternator will benchmark this indirectly via perf_alternator, with auth globally enabled now (btw forgot to add auth params to it, need to fix).

After deeper check I think I won't enable alternator auth in this patchset, it requires additional work in perf_alternator to support it, and it's much less used feature.

ptrsmrn commented:
From the PRs description:
> Enables auth in training clusters
> Adds auth connections stress workload

@mykaul we were discussing why adding new workloads is necessary and why not only enable auth in the already existing workloads - well, the thing is, as @nuivall put it, that the currently existing workloads don't rely on auth enough (don't create that many connections), so a separate workload, a one that creates many connections, had to be added in order to expose auth more to the optimizer.

ptrsmrn commented:
> From the PRs description:
> 
> > Enables auth in training clusters
> > Adds auth connections stress workload
> 
> @mykaul we were discussing why adding new workloads is necessary and why not only enable auth in the already existing workloads - well, the thing is, as @nuivall put it, that the currently existing workloads don't rely on auth enough (don't create that many connections), so a separate workload, a one that creates many connections, had to be added in order to expose auth more to the optimizer.

@michoecho can adding a new workload (and optimize scylla differently because of that) negatively impact the current performance for other workloads?

mykaul commented:
The results for counters are really interesting. I wonder what else we could be missing that could benefit from PGO. @avikivity - how about the tablets load balancer code?
The results for auth are probably a bit disappointing, but I wasn't expecting much from them in the 1st place. It is interesting that even with an auth-heavy workload we don't see any difference though.

nit: I don't see the reason to backport any of this. It's not a bug fix.

ptrsmrn commented:
> by enabling authentication

@nuivall Can you please extend this by adding "by enabling authentication andauthorization" please?

ptrsmrn commented on pgo/pgo.py:
```diff
@@ -514,7 +516,7 @@ def cs_command(cmd: list[str], n: int, node: str, cl: str, pop: Optional[str] =
     ] + (["no-warmup"] if not warmup else []) + [
     ] + (["-pop", pop] if pop else []) + [
-        "-mode", "native", "cql3", "protocolVersion=4",
+        "-mode", "native", "cql3", "protocolVersion=4", "user=cassandra", "password=cassandra",
```
I wonder if using a custom non-superuser wouldn't be better, e.g. there can be more checks done in the code (so more instructions) when we're validating permissions of a non-superuser. OTOH it seems like c-s requires a superuser to work, so that wouldn't be easy to change.

nuivall replied:
I don't think we need to execute every feature in every workload. It's the same as in others, we don't have single huge workload with repair,  counters, and so on but we split it.

nuivall commented on pgo/auth_conns_stress.py:
```diff
@@ -0,0 +1,62 @@
+    cluster = Cluster([host], auth_provider=auth_provider)
+    try:
+        # Prepare once (needs a session). Then reuse prepared statement with new sessions per iteration.
+        prep_session = cluster.connect()
```
ok

nuivall commented on pgo/auth_conns_stress.py:
```diff
@@ -0,0 +1,62 @@
+    try:
+        # Prepare once (needs a session). Then reuse prepared statement with new sessions per iteration.
+        prep_session = cluster.connect()
+        prep_session.set_keyspace('auth_test')
+        prepared = prep_session.prepare("SELECT v FROM auth_table WHERE pk=?")
```
ok

nuivall commented on pgo/auth_conns_stress.py:
```diff
@@ -0,0 +1,62 @@
+    finally:
+        cluster.shutdown()
+
+with concurrent.futures.ThreadPoolExecutor(max_workers=concurrency) as ex:
```
> Maybe that'll descrease the execution time of this script from a couple of minutes to under a minute

This statement is based on what?

I don't think we care much how long training process takes, until it's below 30m-1h

nuivall commented on pgo/exec_cql.py:
```diff
@@ -0,0 +1,76 @@
+Execute CQL statements from a file where each non-empty, non-comment line is exactly one CQL statement.
+Requires python cassandra-driver. Stops at first failure.
+Usage:
+  ./exec_cql.py --file pgo/conf/counters.yaml [--host 127.0.0.1 --port 9042]
```
ok

ptrsmrn commented on pgo/exec_cql.py:
```diff
@@ -0,0 +1,76 @@
+            line = raw.strip()
+            if not line or line.startswith('--'):
+                continue
+            if line.endswith(';'):
+                line = line[:-1].strip()
```
Seems redundant

ptrsmrn commented on pgo/exec_cql.py:
```diff
@@ -0,0 +1,76 @@
+from typing import Sequence
+
+def read_statements(path: str) -> list[tuple[int, str]]:
+    sts: list[tuple[int, str]] = []
```
sts -> stms maybe?

ptrsmrn commented:
> > > Adds auth connections stress workload
> > > Introduces a workload that uses derived roles and permissions, stressing auth
> > > code paths while also creating a new connection per request to exercise
> > > server transport handling.
> > 
> > 
> > Is this really a hot path? IIRC the derived permissions are cached, so the code to calculate derived permissions isn't touched often.
> 
> Yes I think in both cases code paths are not very big, but maybe the cache part can be optimized too. There is also the whole connection management, startup frames parsing, semaphores etc. I don't know how well this optimizes. I'm trying to test before/after, although we don't have full connection storm perf test.

You can try using @andrzej-jackowski-scylladb 's script reproducing a connection storm from his PR https://github.com/scylladb/scylladb/pull/25412

ptrsmrn commented on pgo/auth_conns_stress.py:
```diff
@@ -0,0 +1,62 @@
+rem = iterations % concurrency
+counts = [base + (1 if i < rem else 0) for i in range(concurrency)]
+
+def worker(n: int):
```
> n is amount for work

OK, but what is this work? How many CQL request this worker needs to execute?

michoecho replied:
The `n` in `worker(n)` clearly stands for the ID of the worker.

michoecho commented:
> Performance results for auth PGO there seems to be no difference, or to small to measure:
>
> scylladb pgo_auth ≡ ◦ ⤖ python3 ./pgo/auth_conns_stress.py localhost cassandra cassandra 10000 100 &
> scylladb pgo_auth ≡ ◦ ⤖ perf stat -e instructions --timeout 5000 -p 51591
> 
> on both before and after instructions counter varies from 179,818,558,011 to 180,664,528,198.

@nuivall You chose a bad measurement. Consider:

Program A: 10k instructions per request, 300k requests per second.
Program B: 20k instructions per request, 600k requests per second.

Both programs execute the same number of instructions per 5 seconds. But program B is 2x faster.

Also: did you make sure that Scylla is the bottleneck in this test?

> @michoecho can adding a new workload (and optimize scylla differently because of that) negatively impact the current performance for other workloads?

@ptrsmrn Yes, it can. From my limited experience, it's not that easy to pessimize other workloads by adding new ones, but it's possible.

nuivall commented:
> > Performance results for auth PGO there seems to be no difference, or to small to measure:
> > scylladb pgo_auth ≡ ◦ ⤖ python3 ./pgo/auth_conns_stress.py localhost cassandra cassandra 10000 100 &
> > scylladb pgo_auth ≡ ◦ ⤖ perf stat -e instructions --timeout 5000 -p 51591
> > on both before and after instructions counter varies from 179,818,558,011 to 180,664,528,198.
> 
> @nuivall You chose a bad measurement. Consider:
> 
> Program A: 20k instructions per request, 300k requests per second. Program B: 10k instructions per request, 600k requests per second.
> 
> Both programs execute the same number of instructions per 5 seconds. But program B is 2x faster.
> 
> Also: did you make sure that Scylla is the bottleneck in this test?

Yes you are right, that simplification was due to lack of existing test but it went too far.

nuivall commented:
v3:
- addressed comments
- added os.fork in python stress to make sure concurrency is maximal
- removed cluster.shutdown it causes problems while testing overload and we don't need it as we close the process anyway

Still don't see any improvement for auth, this time I measured time it takes to execute 30k successful requests, it's not accurate metric but quite stable, based on this I think improvement during stress scenario has to be below 3% (if any).

avikivity commented on pgo/exec_cql.py:
```diff
@@ -0,0 +1,74 @@
+"""exec_cql.py
+Execute CQL statements from a file where each non-empty, non-comment line is exactly one CQL statement.
+Requires python cassandra-driver. Stops at first failure.
+Usage:
```
Ah, I see you want to run it as part of training.

michoecho replied:
I think this is just `cqlsh -f`, but @nuivall added this to avoid a dependency on cqlsh.

avikivity commented on pgo/pgo.py:
```diff
@@ -661,6 +661,27 @@ async def train_decommission(executable: PathLike, workdir: PathLike) -> None:
+        reader_pass = "password1"
+        # run 30k connection requests
+        await run_checked(["python3", "./auth_conns_stress.py", addrs[0], reader_user, reader_pass, "30", "10", "100"])
+    await merge_profraw(workdir)
```
Normally I'd be very unhappy at forking 30k processes, but here I guess the goal is to simulate client startup, so it's sensible.

michoecho replied:
I think we're forking 30 processes, with 10 threads each.

nuivall replied:
I fork 30. Each is 10 threads (this is mostly done as I was fiddling with locally).

I will add named arguments to the script to make it more readable as Michal mentioned below.

nuivall commented on pgo/auth_conns_stress.py:
```diff
@@ -0,0 +1,116 @@
+    sys.exit(1)
+
+def prepare_select(cluster):
+    # Prepare once (needs a session). Retry silently on transient NoHostAvailable, it's most likely an overload.
+    while True:
```
for the scripted scenario it's probably fine to not retry here, but when I played with different parameters you may need retries when e.g. you spawn lots of processes small requests per process number

I'll add log line

nuivall commented on pgo/auth_conns_stress.py:
```diff
@@ -0,0 +1,116 @@
+
+Process + thread fan-out stress tool creating a NEW session per request.
+
+Args (order on CLI):
```
ok

nuivall commented on pgo/pgo.py:
```diff
@@ -659,6 +661,27 @@ async def train_decommission(executable: PathLike, workdir: PathLike) -> None:
+        reader_user = "reader_role"
+        reader_pass = "password1"
+        # run 30k connection requests
+        await run_checked(["python3", "./auth_conns_stress.py", addrs[0], reader_user, reader_pass, "30", "10", "100"])
```
ok
